website/source/docs/auth/gcp.html.md

```diff
@@ -0,0 +1,339 @@
+
+The Vault authentication workflow for IAM service accounts is as follows (see diagram):
+
+  1. A client with IAM service account credentials generates a signed JWT using the IAM [projects.serviceAccounts.signJwt](https://cloud.google.com/iam/reference/rest/v1/projects.serviceAccounts/signJwt) method. See [usage](#iam-authentication-token) for the expected format and example code.
+  2. The client sends this JWT to Vault in a login request with a role name. This role should have type `iam`
+  3. Vault grabs the `kid` header value, which contains the ID of the key-pair used to generate the JWT, and the `sub` ID/email to find the service account key. If the service account does not exist or the key is not linked to the service account, Vault will deny authentication.
+  4. Vault authorizes the confirmed service account against the given role. See [authorization section](#authorization) to see how each type of role handles authorization.
+
+## Authorization
+For `gcp`, login is per-role. Each role has a specific set of restrictions that
+an authorized entity must fit in order to login. This restrictions are specific
+to the role type.
+
+Currently supported role types are: `iam`
+
+Vault validates an authenticated entity against the role and uses the role to
+determine information about the lease, including Vault policies assigned and
+TTLs. For a full list of accepted arguments, see [role path usage](#roles).
+
+### `iam` Roles
+
+`iam` roles support the following role restrictions:
+
+  * `service_accounts`: A list of service accounts that are allowed to login as
+    this role. This list accepts either service account IDs or emails.
+
+**Note**: In the future we hope to support restrictions related to
+IAM policies/permissions. With current IAM APIs, querying all IAM roles or permissions
+(set or inherited) on an entity is very difficult to do and overall ends with not a
+great story. We plan to wait until more functionality is supported before
+attempting to add permission-based constraints.
+
+## Usage
+### Config
+```
+vault write auth/gcp/config \
+  credentials=...
+```
+Parameters:
+
+  * **'credentials'**: The contents of a Google credentials JSON file to be
+    used by Vault to make calls to GCP APIs. If you would rather specify the
+    file, you can use `credentials="@path/to/creds.json`.
+
+    If this value is not specified or if it is explicitly set to empty, Vault will attempt to use [Application Default Credentials](https://developers.google.com/identity/protocols/application-default-credentials) for that server's machine.
+
+The GCP permissions Vault currently requires are:
+
+  * `iam.serviceAccounts.get`
+  * `iam.serviceAccountKeys.get`
+
+
+### Roles
+
+```sh
+vault write auth/gcp/role/myrole \
+  type='iam' \
+  ...
+
+vault read auth/gcp/role/myrole
+
+vault list auth/gcp/role
+```
+
+Parameters:
+
+  * `type`: Required. Type of the role. Currently supported: `iam`.
```

Comment:
The pattern in this page is significantly different than the documentations followed for other authentication backends. The usual pattern is to only quote working samples in this page and to provide link to the API page to refer to the detailed list of parameters and their meanings. You might want to remove all the redundancies. In the interest of time, feel free to defer this task until after the release if you are occupied with other things.